### DISCUSSION TITLE
还是没理解MVVM该用在哪里

### DISCUSSION THREAD

**Discussion by Hakoyu** [BYJRK/DotNet-Discussions#11] [MVVM]
到头来就在项目中加了几个 `[ObservableProperty]` 少写几行代码咯

**Comment by BYJRK** [comment]
MVVM 是一个很庞大的话题，可以说是一种全新的界面设计理念，不太好几句话讲清楚，我之后有打算专门出几期视频聊一聊我对 MVVM 的理解。目前简单说一下的话，MVVM 的主要目的是为了前后端的解耦，这样的好处是后端（ViewModel）可以写单元测试，界面逻辑放在后端，使得前端换框架更为方便，以及配套使用依赖注入（DI）等等。

具体举个简单的例子，比如有用户名和密码两个文本框以及一个注册按钮，界面逻辑包括：

1. 两个文本框都不为空的时候，按钮可用，否则不可用
2. 姓名文本框会提示你姓名是否合法（如果不合法，文本框边框会变色）
3. 密码文本框会提示你密码是否有效（包括包含字符和长度等）

如果用传统的 MVC，那么你一定会需要去注册 `TextBox` 的 `TextChanged` 事件等，并在事件订阅的方法中去修改按钮的 `IsEnabled` 属性以及文本框的 `BorderBrush` 属性等；
如果用 MVVM，那么你需要做的是在 ViewModel 中进行相关的逻辑判断，并借助 Validation 等来实现内容校验，然后前端与后端的属性进行绑定，并添加 `DataTrigger` 等来实现对界面效果的实现

**Comment by Hakoyu** [comment]
通俗来讲就是将View的东西绑定到ViewModel,这样就不用去调用View本身的属性
从效果上来说并不是很好,甚至有些复杂,传统方法中遇到的问题在MVVM中同样会遇到,也没有对多线程的额外支持
而且很多时候都是传统事件和ViewModel事件并存
例如列表中的项目太多 就得用异步加载
```csharp
// DataGrid.ItemsSource = items;
foreach (var item in items)
    Dispatcher.InvokeAsync(() => DataGrid.Items.Add(item), DispatcherPriority.Background);
```
在MVVM中并没有对这种加载提供便利,绑定的数据依旧得一个一个填入

**Reply by BYJRK** [reply]
你的这个做法是不符合 MVVM 模式的，因为你的 `Items.Add` 实际上是在直接操作界面控件。MVVM 典型做法是 ViewModel 里面有一个 `ObservableCollection`，然后前台的 `DataGrid` 的 `ItemsSource` 属性绑定这个 `Collection`，然后在 VM 中对 `Collection` 进行增删等操作，并实时同步到前台

**Reply by Hakoyu** [reply]
我之前看过你视频实现的DataGrid筛选
```csharp
public partial class ViewModel : ObservableObject
{
    [ObservableProperty]
    ICollectionView? collectionView;
    [ObservableProperty]
    string? filterText;
    public string filterType = strName;
    partial void OnFilterTextChanged(string? value) => CollectionView?.Refresh();
    public ViewModel(IEnumerable<ModShowInfo> modShowInfos)
    {
        ChangeCollectionView(modShowInfos);
    }
    public void ChangeCollectionView(IEnumerable<ModShowInfo> modShowInfos)
    {
        CollectionView = CollectionViewSource.GetDefaultView(modShowInfos);
        CollectionView.Filter = (o) =>
        {
            if (string.IsNullOrEmpty(filterText))
                return true;
            if (o is not ModShowInfo showInfo)
                return true;
            return filterType switch
            {
                strName => showInfo.Name.Contains(filterText, StringComparison.OrdinalIgnoreCase),
                strId => showInfo.Id.Contains(filterText, StringComparison.OrdinalIgnoreCase),
                strAuthor => showInfo.Author.Contains(filterText, StringComparison.OrdinalIgnoreCase),
                strUserDescription => showInfo.UserDescription.Contains(filterText, StringComparison.OrdinalIgnoreCase),
                _ => throw new NotImplementedException()
            };
        };
    }
}
```
但实际上并没有解决痛点

使用`CollectionView.Refresh()`?
拜托 `DataGrid`性能实在是捉鸡
还是老老实实绑定一个`ObservableCollection`然后给`ViewModel`加个`Add`和`Remove`比较靠谱

**Reply by eijjxx** [reply]
我感觉这个应该是虚拟化解决的问题吧？
试过虚拟化吗？
```c#
<DataGrid  VirtualizingPanel.IsVirtualizing="True" 
                   VirtualizingPanel.IsContainerVirtualizable="True"
                   VirtualizingPanel.VirtualizationMode="Recycling"/>
```
从MVVM的概念上而言，这应该是View和View-Controller解决的问题，不该在ViewModel端解决

**Reply by kumamiko** [reply]
> 我之前看过你视频实现的DataGrid筛选
> ```csharp
> public partial class ViewModel : ObservableObject
> {
>     [ObservableProperty]
>     ICollectionView? collectionView;
>     [ObservableProperty]
>     string? filterText;
>     public string filterType = strName;
>     partial void OnFilterTextChanged(string? value) => CollectionView?.Refresh();
>     public ViewModel(IEnumerable<ModShowInfo> modShowInfos)
>     {
>         ChangeCollectionView(modShowInfos);
>     }
>     public void ChangeCollectionView(IEnumerable<ModShowInfo> modShowInfos)
>     {
>         CollectionView = CollectionViewSource.GetDefaultView(modShowInfos);
>         CollectionView.Filter = (o) =>
>         {
>             if (string.IsNullOrEmpty(filterText))
>                 return true;
>             if (o is not ModShowInfo showInfo)
>                 return true;
>             return filterType switch
>             {
>                 strName => showInfo.Name.Contains(filterText, StringComparison.OrdinalIgnoreCase),
>                 strId => showInfo.Id.Contains(filterText, StringComparison.OrdinalIgnoreCase),
>                 strAuthor => showInfo.Author.Contains(filterText, StringComparison.OrdinalIgnoreCase),
>                 strUserDescription => showInfo.UserDescription.Contains(filterText, StringComparison.OrdinalIgnoreCase),
>                 _ => throw new NotImplementedException()
>             };
>         };
>     }
> }
> ```
> 但实际上并没有解决痛点
> 
> 使用`CollectionView.Refresh()`?
> 拜托 `DataGrid`性能实在是捉鸡
> 还是老老实实绑定一个`ObservableCollection`然后给`ViewModel`加个`Add`和`Remove`比较靠谱

这个filterText频繁修改，稍微多点东西，界面会卡到怀疑人生（啥原因不太懂，可能是多次触发事件，在主线程竞争阻塞了渲染？）
这种实时输入筛选的功能，建议用ReactiveUI，简直爽歪歪。

**Comment by Hakoyu** [comment]
像是后台添加控件 比如说读取目录中的图片并显示
难道是在`ViewModel`中读取文件然后传递给`View`绑定吗
如果要实现动态增删,使用`ObservableCollection<Image>`进行存储,那我还是拿到了前台控件

**Reply by BYJRK** [reply]
没错，你在 VM 里面读取文件，然后添加到一个 `Collection` 中，并让 View 中的控件绑定；
至于图片，你不应该在 ViewModel 中出现 View 的控件类型，所以你应该存储的是图片的路径（`Uri`）。如果某些前端框架不能直接把 Image 的 `Source` 属性绑定到字符串（或 `Uri`），那么你还可以写一个 `ValueConverter`。

**Comment by Hakoyu** [comment]
感觉会把很多东西复杂化
比如单页切换,MVC只要个`Dic<string,Page>`,然后看`Frame`要啥直接放过去就行
在 VM 中不能存`Page`顶多存个`List<string> pageNames`,切换页面这种事情只能交给 View 层
如果要实现`ListBox`切换页面...我觉得是自讨苦吃

**Reply by BYJRK** [reply]
你可能还是没搞清楚 MVVM 到底是为了解决一个什么问题，总觉得是在各种给自己找麻烦，因为刚从 MVC 转到 MVVM 确实会频繁发现一些很熟悉的操作变成了“错误”的，需要去做出各种各样的改变。但其实不是这个样子的，只能等待你自己之后的发掘了。
至于单页切换，有很多常见的 MVVM 实现方式，比如使用 `ValueConverter`，使用 `ViewModelLocator`，使用 IoC，或者附加属性和行为等

**Reply by Hakoyu** [reply]
我在将我的项目修改为MVVM模式 https://github.com/Hakoyu/StarsectorTools
但是没有什么资料 搜个mvvm后台添加ListBoxItem绑定Click能有13年的玩意
我现在大部分时间都在搜怎么实现,然后验证搜到的玩意是否能用

**Reply by BYJRK** [reply]
我之后会逐渐分享更多的有关WPF MVVM的内容。你如果有相关的问题，也欢迎在这里提出

**Reply by eijjxx** [reply]
MVVM在解耦、测试、修改、多人合作方面有很大的优势
代码量是一定会提升的，结构也一定会复杂（对比写的很好的不用mvvm的代码）

对于单页的功能简单的代码真没必要用
对于复杂的代码可以说是不可或缺

CommunityToolkit.mvvm啊，prism啊，只是用来标准和简化mvvm的代码实现

**Comment by notplay3** [comment]
richtextbox怎么做到用mvvm实现绑定？

**Comment by yijie0110** [comment]
MVVM最大的好处就是对View和ViewModel进行解耦, 这个过程要做出一些牺牲以及学习跟多的知识来应对"牺牲".要知道mvvm不是只为了给wpf用的,当你严格的执行了mvvm的标准后,那么你只要你把ViewModel的代码独立出去,然后你可以直接对接上新的ui框架(这里可以是网页比如.net的blazor, 多平台的Maui,Avalonia,Xamarin等等)而你只需要重写ui不太需要关心业务逻辑了,还有就是解耦后VM写测试就很方便了.
根据以上逻辑,当你在写VM的时候想一下"如果我把ui(Wpf)换掉了 我这viewModel的代码要不要改?"如果不用那我觉得你已经在遵守了mvvm的框架规则了,说白了就是VM里不引用任何有关ui框架的库
所以MVVM是一个很大的东西. 如果只是直接做一些小工具不考虑更换UI框架或者多平台发布,就没必要特意追求符合mvvm的规则.
为了符合mvvm有的时候写UI确实挺绕的.掉头发啊!

**Comment by meditator10** [comment]
> 到头来就在项目中加了几个 `[ObservableProperty]` 少写几行代码咯

单看你这个原始问题描述，不是想问MVVM框架本身吧
是不是想问MVVMToolkit与WPF MVVM之间版本迭代所带来的诸如[ObservableProperty]特性？这些确实是源生成器帮你写好了关联的字段和属性，省去了一些代码。除此之外还有很多如前面所述，依赖注入、Messager、[RelayCommand]等等
